lib/definition/model.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+module Definition
+  class InvalidModelError < StandardError
+    attr_accessor :conform_result
+
+    def initialize(conform_result)
+      super(conform_result.error_message)
+      self.conform_result = conform_result
+    end
+  end
+
+  class Model
+    class << self
+      def conform(value)
+        _definition.conform(value)
+      end
+
+      def required(key, definition)
+        _define_attr_accessor(key)
+        _definition.required(key, definition)
+      end
+
+      def optional(key, definition, **opts)
+        _define_attr_accessor(key)
+        _definition.optional(key, definition, **opts)
+      end
+
+      def option(option_name)
+        _definition.option(option_name)
+      end
+
+      def _define_attr_accessor(key)
+        define_method(key) do
+          @_attributes.fetch(key, nil)
+        end
+      end
+
+      def _definition
+        @_definition ||= ::Definition.Keys {}
+      end
+    end
+
+    def initialize(hash = nil, **kwargs)
+      result = self.class.conform(hash || kwargs)
+      raise InvalidModelError.new(result) unless result.passed?
+
+      @_attributes = result.value.freeze
+    end
+
+    def new(**kwargs)
+      self.class.new(**to_h.merge!(kwargs))
+    end
+
+    def ==(other)
+      @_attributes == other.instance_variable_get(:@_attributes)
+    end
+
+    def to_h
```

Comment:
i changed this now to deeply convert object to hashes so that if you nest definition models, that those are also converted to hashes. This was also something that proved to be needed in the activity_file_import service where the repository strategies gem just converts the model into a database document by calling `.to_h` on it